packages/flutter/lib/src/widgets/implicit_animations.dart

```diff
@@ -565,6 +579,243 @@ abstract class AnimatedWidgetBaseState<T extends ImplicitlyAnimatedWidget> exten
+  /// it can be used in a `const` constructor!
+  final LerpCallback<T> lerp;
+
+  /// Called each time the animation completes.
+  ///
+  /// This can be useful to trigger additional actions (e.g. another animation)
+  /// at the end of the current animation.
+  final VoidCallback? onEnd;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget? child;
+
+  /// Called each frame to build a widget, given the current state
+  /// of the value's animation.
+  @protected
+  Widget build(BuildContext context, T value);
+
+  @override
+  State<AnimatedValue<T>> createState() => _AnimatedValueState<T>();
+}
+
+class _AnimatedValueState<T extends Object> extends State<AnimatedValue<T>>
+    with SingleTickerProviderStateMixin {
+  late T from = widget.initialValue ?? widget.value;
+  late T target = widget.value;
+  late T value = from;
+
+  Ticker? _ticker;
+
+  void _animate() {
+    final T newTarget = widget.value;
+    if (mounted && newTarget != value) {
+      _ticker?.stop();
+
+      if (widget.duration == Duration.zero) {
+        widget.onEnd?.call();
+        value = newTarget;
+      } else {
+        from = value;
+        target = newTarget;
+        value = widget.lerp(from, target, 0)!;
+        _ticker?.start();
+      }
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _ticker = createTicker((Duration elapsed) {
+      if (value == target) {
+        widget.onEnd?.call();
+        return _ticker!.stop();
+      }
+
+      final double progress = elapsed.inMicroseconds / widget.duration.inMicroseconds;
+      if (progress >= 1.0) {
+        widget.onEnd?.call();
+        _ticker!.stop();
+        if (mounted) {
+          setState(() {
+            value = target;
+          });
+        }
+        return;
+      }
+
+      if (mounted) {
+        final double t = widget.curve.transform(math.max(progress, 0.0));
+        setState(() {
+          value = widget.lerp(from, target, t)!;
+        });
+      } else {
+        _ticker!.stop();
+      }
+    });
+    _animate();
+  }
+
+  @override
+  void didUpdateWidget(AnimatedValue<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _animate();
+  }
+
+  @override
+  void dispose() {
+    _ticker?.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) => widget.build(context, value);
+}
+
+class _AnimatedValueBuilder<T extends Object> extends AnimatedValue<T> {
+  const _AnimatedValueBuilder(
+    super.value, {
+    super.key,
+    super.initialValue,
+    required super.duration,
+    super.curve,
+    required super.lerp,
+    super.onEnd,
+    required this.builder,
+    super.child,
+  });
+
+  /// {@macro flutter.widgets.ValueWidgetBuilder}
+  final ValueWidgetBuilder<T> builder;
+
+  @override
+  Widget build(BuildContext context, T value) => builder(context, value, child);
+}
+
+double? _lerpDouble(double? a, double? b, double t) {
```

Comment:
Can this just use https://main-api.flutter.dev/flutter/dart-ui/lerpDouble.html?

Reply:
That's what I did initially, but I found out that returning `b` by default is required in order to avoid breaking a few tests.

Reply:
This should probably be documented. Although, I find it strange that this matters for tests. What are the test failures you're seeing?